python/lsst/pipe/base/graphBuilder.py

```diff
@@ -501,6 +501,9 @@ def __init__(self, registry: Registry, run: str):
     def resolveRef(self, ref: DatasetRef) -> DatasetRef:
         if ref.id is not None:
             return ref
+        # For components we need their parent dataset type.
+        if ref.isComponent():
+            ref = ref.makeCompositeRef()
```

Comment:
If we're calling this method in an order consistent with the graph, a component appearing here should mean the lookup in `self.resolved` should always succeed.  It might be worth using an else clause here to ensure that; we certainly don't want some other bug to lead to generating random UUIDs for components.

Reply:
If a component is a pure input (not an intermediate) then its parent should not be in `self.resolved`? But then it should be already resolved, which I can certainly check here too.